SwerveModule and DriveSubsystem

tails618 opened the pull request:
- Create DriveSubsystem and SwerveModule (and relevant constants/calibrations) based on 2023
- Convert drive motor from SparkMAX to TalonFX
- Updated `PathToPoint`, `PoseToPath`, and `followTrajectoryCommand` to PathPlanner 2024 API
- Added GeometryUtils

tails618 committed:
Added REVLib

tails618 committed:
Added utils and global constants

tails618 committed:
Added SwerveModule and Drive Constants/Cals

tails618 committed:
Some progress on Swerve + phoenix6

Panda8188 committed:
Unfinished Drive Subystem

* Added GeometryUtils.java
* Added (empty) Lights.java
* Added can Id's to robot map and placeholder variables to constants
* DriveSubsystem.java is unfinished

tails618 committed:
Lots of drive stuff

avakimc committed:
Filled In followTrajectoryCommand Method

- added PathPlanner vendor libraries
- filled in followTrajectoryCommand method based on last year's code
- HOPEFULLY replaced the old PathPlanner methods with the correct ones 🤞
- added necessary parameter methods for the followTrajectoryCommand method

avakimc committed:
Misc Things Missing From Previous Commit

- DriveConstants.DRIVE_BASE_RADIUS_METERS
- comment

avakimc committed:
Eleanor/Allison/Sofia/Ava Updating PathToPoint and PoseToPath Methods for New PathPlanner API

voidSevenSevenSix committed:
updated swervemodule to use talonfx for the driving motor

balghane commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -4,16 +4,55 @@
+    public static final boolean GYRO_REVERSED = false;
+  }
+
+  public final class SwerveSubsystem {
+    public static final double IMU_PITCH_BIAS_DEG = PLACEHOLDER_DOUBLE;
+  }
```
No balancing so no pitch stuff is needed

balghane commented on src/main/java/frc/robot/RobotMap.java:
```diff
@@ -0,0 +1,16 @@
+
+public class RobotMap {
+  public static final int PIGEON_CAN_ID = Constants.PLACEHOLDER_INT;
+    /** Driving SPARK MAX CAN IDs */
```
Talon not spark

balghane commented on src/main/java/frc/robot/subsystems/drive/DriveCal.java:
```diff
@@ -0,0 +1,61 @@
+public class DriveCal {
+    public static final int SPARK_INIT_RETRY_ATTEMPTS = 5;
+
+    /** Input meters/second, output [-1,1] */
```
Are the units for this the same as the units were in Rev-world?

tails618 replied:
`TalonFX.setControl` (L190 in SwerveModule.java) takes an input [-1,1]; I *think* that's the place to look, in which case this is correct?

balghane commented on src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java:
```diff
@@ -0,0 +1,617 @@
+            this::getPose, // pose supplier
+            this::getChassisSpeeds, // robot relative chassis speeds supplier
+            this::setOutputRobotRelativeSpeeds, // robot relative chassis speeds consumer
+            new PIDConstants(DriveCal.DRIVING_P, DriveCal.DRIVING_I, DriveCal.DRIVING_D), // PID Constants for
+                                                                                          // translation
+            new PIDConstants(DriveCal.TURNING_P, DriveCal.TURNING_I, DriveCal.TURNING_D), // PID Constants for rotation
```
I think the path planner PIDs for total robot translation and total robot rotation were different than the module drive and module rotation PIDs

tails618 replied:
@avakimc

balghane commented on src/main/java/frc/robot/subsystems/drive/DriveConstants.java:
```diff
@@ -0,0 +1,64 @@
+     */
+    public static final boolean TURNING_ENCODER_INVERTED = true;
+
+    /** Multiplier for wheel diameter based on empirical on-field measurement */
+    // First number: adjustment for midwest
+    // Second number: adjustment from midwest to practice field
+    // Third number: adjustment from practice field to wisconsin
+    // Fourth number: Wisconsin to Buckeye
+    public static final double WHEEL_DIAMETER_FUDGE_FACTOR = 0.898 * 1.120 * 0.992 * 1.035;
```
Let's reset this to 1?

tails618 replied:
i should've caught this one 😅

balghane commented on src/main/java/frc/robot/subsystems/drive/DriveConstants.java:
```diff
@@ -0,0 +1,64 @@
+    /** the distance from the center of the robot to the furthest module */
+    public static final double DRIVE_BASE_RADIUS_METERS = Constants.PLACEHOLDER_DOUBLE;
+
+    public static final double DRIVING_MOTOR_REDUCTION = 4.8;
```
placeholder

balghane commented on src/main/java/frc/robot/subsystems/drive/DriveConstants.java:
```diff
@@ -0,0 +1,64 @@
+import edu.wpi.first.math.util.Units;
+import frc.robot.Constants;
+
+public class DriveConstants {
```
Why are some things in Constants and other things in DriveConstants?

tails618 replied:
Two different people did it different ways and I'll refrain from naming the person that wasn't me

balghane commented on src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java:
```diff
@@ -0,0 +1,617 @@
+  private ChassisSpeeds lastSetChassisSpeeds = new ChassisSpeeds(0.0, 0.0, 0.0);
+  public Optional<Pose2d> targetPose = Optional.empty();
+  /** include path planner initialization here */
+  private MedianFilter pitchFilter = new MedianFilter(Constants.PLACEHOLDER_INT);
+  private double latestFilteredPitchDeg = Constants.PLACEHOLDER_DOUBLE;
```
prob remove

balghane commented on src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java:
```diff
@@ -0,0 +1,617 @@
+  private final Pigeon2 gyro = new Pigeon2(RobotMap.PIGEON_CAN_ID);
+  private ChassisSpeeds lastSetChassisSpeeds = new ChassisSpeeds(0.0, 0.0, 0.0);
+  public Optional<Pose2d> targetPose = Optional.empty();
+  /** include path planner initialization here */
```
?

balghane commented on src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java:
```diff
@@ -0,0 +1,617 @@
+  public DriveSubsystem(Lights lightsSubsystem, BooleanSupplier isTimedMatchFunc) {
+    /** Need to factory default settings for gyro, but no function exists */
+    gyro.reset();
+    /** Need to configure a mount pose for gyro, but no function exist */
```
ConfigMountPose

balghane commented on src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java:
```diff
@@ -0,0 +1,617 @@
+  private BooleanSupplier isTimedMatch;
+
+  public DriveSubsystem(Lights lightsSubsystem, BooleanSupplier isTimedMatchFunc) {
+    /** Need to factory default settings for gyro, but no function exists */
```
ConfigAllSettings but I don't think we actually need to set anything

balghane commented on src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java:
```diff
@@ -0,0 +1,617 @@
+    this.isTimedMatch = isTimedMatchFunc;
+  }
+
+  public double getFilteredPitch() {
+    return latestFilteredPitchDeg - Constants.SwerveSubsystem.IMU_PITCH_BIAS_DEG;
+  }
```
prob remove

balghane commented on src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java:
```diff
@@ -0,0 +1,617 @@
+    frontRight.periodic();
+    rearLeft.periodic();
+    rearRight.periodic();
+    odometry.update(Rotation2d.fromDegrees(gyro.getYaw().getValue()), getModulePositions());
```
Let's throw in a TODO to consider moving this to a thread that loops faster

balghane commented on src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java:
```diff
@@ -0,0 +1,617 @@
+   * applies the robot-relative output speeds of the FollowPathHolonomic Command
+   * adapted from drive() (but doesn't need the code to turn xSpeed, ySpeed, rot,
+   * fieldRelative into a ChassisSpeeds object)
+   * does not call the correctForDynamics method because we pass in a corrected
+   * lastSetChassisSpeeds (// TODO is this right???)
```
I doubt path planner is doing this correction

tails618 replied:
@avakimc

balghane commented on src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java:
```diff
@@ -0,0 +1,617 @@
+   * https://www.chiefdelphi.com/t/whitepaper-swerve-drive-skew-and-second-order-kinematics/416964
+   */
+  private static ChassisSpeeds correctForDynamics(ChassisSpeeds originalSpeeds) {
+    final double LOOP_TIME_S = 0.02;
```
There was some discussion about arbitrarily making this larger helping, can we throw in a TODO to test that?

balghane commented on src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java:
```diff
@@ -0,0 +1,617 @@
+   *                      field.
+   */
+  public void drive(double xSpeed, double ySpeed, double rot, boolean fieldRelative) {
+    if (isTimedMatch.getAsBoolean()
+        && DriverStation.isTeleop()
+        && DriverStation.getMatchTime() < 0.3) {
+      setX();
+      return;
+    }
```
I don't think we need this anymore, this was specific to balancing

balghane commented on src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java:
```diff
@@ -0,0 +1,617 @@
+   */
+  public double getTurnRate() {
+    return gyro.getRate()
+        * (Constants.SwerveDrive.GYRO_REVERSED ? Constants.PLACEHOLDER_DOUBLE : Constants.PLACEHOLDER_DOUBLE);
```
Why are these placeholders?

balghane commented on src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java:
```diff
@@ -0,0 +1,617 @@
+    }
+  }
+
+  public void setForward() {
+    frontLeft.setDesiredState(new SwerveModuleState(0, Rotation2d.fromDegrees(0)));
+    frontRight.setDesiredState(new SwerveModuleState(0, Rotation2d.fromDegrees(0)));
+    rearLeft.setDesiredState(new SwerveModuleState(0, Rotation2d.fromDegrees(0)));
+    rearRight.setDesiredState(new SwerveModuleState(0, Rotation2d.fromDegrees(0)));
+  }
```
I think we stopped using this

balghane commented on src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java:
```diff
@@ -0,0 +1,617 @@
+        new FollowPathHolonomic( // old method from last year: PPSwerveControllerCommand
+            path,
+            this::getPose, // pose supplier
+            this::getChassisSpeeds, // robot relative chassis speeds supplier
```
Should we be getting the actual module speeds and converting to a chassis speeds here? Instead of just using the last set point?

tails618 replied:
@avakimc

balghane commented on src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java:
```diff
@@ -0,0 +1,617 @@
+            new PIDConstants(DriveCal.TURNING_P, DriveCal.TURNING_I, DriveCal.TURNING_D), // PID Constants for rotation
+            DriveConstants.DRIVE_WHEEL_FREE_SPEED_METERS_PER_SECOND, // maxModuleSpeed
+            DriveConstants.DRIVE_BASE_RADIUS_METERS,
+            new ReplanningConfig(), // creates a path replanning configuration with the default config
```
Let's throw in a TODO for this

balghane commented on src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java:
```diff
@@ -0,0 +1,617 @@
+            DriveConstants.DRIVE_WHEEL_FREE_SPEED_METERS_PER_SECOND, // maxModuleSpeed
+            DriveConstants.DRIVE_BASE_RADIUS_METERS,
+            new ReplanningConfig(), // creates a path replanning configuration with the default config
+            () -> {
+              return false;
+            }, // boolean supplier for shouldFlipPath
```
Is this right for what we decided?

balghane commented on src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java:
```diff
@@ -0,0 +1,617 @@
+    double latencyAdjustmentSec = 0.00;
+
+    Pose2d curPose = getPose();
+    Pose2d pastPose = getPastPose(latencyAdjustmentSec);
```
Let's add a todo to see if we can get this working with the real latency `latencySec` since right now the latency compensation is basically just turned off

balghane commented on src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java:
```diff
@@ -0,0 +1,617 @@
+    Transform2d finalTransform = new Transform2d(finalPose.getTranslation(), finalPose.getRotation());
+    System.out.println(
+        "Trajectory Transform: " + finalTransform.getX() + " " + finalTransform.getY());
+    Rotation2d finalHeading = Rotation2d.fromDegrees(180);
```
Is this used

balghane commented on src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java:
```diff
@@ -0,0 +1,617 @@
+    Rotation2d finalHeading = Rotation2d.fromDegrees(180);
+    Rotation2d finalHolonomicRotation = finalPose.getRotation();
+
+    PathPlannerPath path = PathPlannerPath.fromPathPoints(Arrays.asList(new PathPoint(finalTransform.getTranslation(),
```
`fromPathPoints` is for a path that's already smoothed. I think we need to use a different function for path planner to smooth the path for us

balghane commented on src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java:
```diff
@@ -0,0 +1,617 @@
+
+    Pose2d finalPose = targetPose.get();
+    Transform2d finalTransform = new Transform2d(finalPose.getTranslation(), finalPose.getRotation());
+    Rotation2d finalHeading = Rotation2d.fromDegrees(180);
```
remove

balghane commented on src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java:
```diff
@@ -0,0 +1,617 @@
+    Pose2d finalPose = targetPose.get();
+    Transform2d finalTransform = new Transform2d(finalPose.getTranslation(), finalPose.getRotation());
+    Rotation2d finalHeading = Rotation2d.fromDegrees(180);
+    Rotation2d finalHolonomicRotation = Rotation2d.fromDegrees(0);
```
we may need to question this

balghane commented on src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java:
```diff
@@ -0,0 +1,617 @@
+    Rotation2d finalHeading = Rotation2d.fromDegrees(180);
+    Rotation2d finalHolonomicRotation = Rotation2d.fromDegrees(0);
+
+    PathPlannerPath path = PathPlannerPath.fromPathPoints(Arrays.asList(new PathPoint(finalTransform.getTranslation(),
```
same comment about fromPathPoints as above

balghane commented on src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java:
```diff
@@ -0,0 +1,617 @@
+    targetHeadingDegrees = getHeadingDegrees() + offsetDegrees;
+  }
+
+  public void setZeroTargetHeading() {
+    targetHeadingDegrees = 0.0;
+  }
```
We may need to swap this out for just "pointing forward" which will flip directions depending on color

balghane commented on src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java:
```diff
@@ -0,0 +1,617 @@
+        .withTimeout(timeoutSec);
+  }
+
+  public void zeroFrontLeftAtCurrentPos() {
+    DriveCal.SWERVE_FRONT_LEFT_ANGULAR_OFFSET_RAD = frontLeft.getEncoderAbsPositionRad();
+    System.out.println(
+        "New Zero for Front Left Swerve: "
+            + DriveCal.SWERVE_FRONT_LEFT_ANGULAR_OFFSET_RAD);
+  }
+
+  public void zeroFrontRightAtCurrentPos() {
+    DriveCal.SWERVE_FRONT_RIGHT_ANGULAR_OFFSET_RAD = frontRight.getEncoderAbsPositionRad();
+    System.out.println(
+        "New Zero for Front Right Swerve: "
+            + DriveCal.SWERVE_FRONT_RIGHT_ANGULAR_OFFSET_RAD);
+  }
+
+  public void zeroRearLeftAtCurrentPos() {
+    DriveCal.SWERVE_BACK_LEFT_ANGULAR_OFFSET_RAD = rearLeft.getEncoderAbsPositionRad();
+    System.out.println(
+        "New Zero for Rear Left Swerve: "
+            + DriveCal.SWERVE_BACK_LEFT_ANGULAR_OFFSET_RAD);
+  }
+
+  public void zeroRearRightAtCurrentPos() {
+    DriveCal.SWERVE_BACK_RIGHT_ANGULAR_OFFSET_RAD = rearRight.getEncoderAbsPositionRad();
+    System.out.println(
+        "New Zero for Rear Right Swerve: "
+            + DriveCal.SWERVE_BACK_RIGHT_ANGULAR_OFFSET_RAD);
+  }
```
Did we ever use this stuff? Maybe remove?

balghane commented on src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java:
```diff
@@ -0,0 +1,617 @@
+    // addChild("X Controller", DriveCal.PATH_X_CONTROLLER);
+    // addChild("Y Controller", DriveCal.PATH_Y_CONTROLLER);
+    // addChild("Theta Controller", DriveCal.PATH_THETA_CONTROLLER);
+    addChild("Rotate to target controller", DriveCal.ROTATE_TO_TARGET_PID_CONTROLLER);
+    addChild("Front Right", frontRight);
+    addChild("Front Left", frontLeft);
+    addChild("Rear Right", rearRight);
+    addChild("Rear Left", rearLeft);
```
Let's remember to try getting these to actually work this year

tails618 replied:
*surely*

balghane commented on src/main/java/frc/robot/utils/AbsoluteEncoderChecker.java:
```diff
@@ -0,0 +1,46 @@
+  private int index;
+  private MedianFilter medianFilter;
+  private double lastCalculatedMedian = 0.0;
+  private static final int NUM_READINGS = 20;
```
Can we put a TODO to rethink this whole thing since it was a nice idea but didn't really work

balghane commented on src/main/java/frc/robot/subsystems/drive/SwerveModule.java:
```diff
@@ -0,0 +1,237 @@
+  void initDriveTalon() {
+    TalonFXConfigurator cfg = drivingTalon.getConfigurator();
+    TalonFXConfiguration toApply = new TalonFXConfiguration();
+    toApply.MotorOutput.Inverted = InvertedValue.Clockwise_Positive;
```
Is this right

balghane commented on src/main/java/frc/robot/subsystems/drive/SwerveModule.java:
```diff
@@ -0,0 +1,237 @@
+            drivingPidTmp.setOutputRange(
+                DriveCal.DRIVING_MIN_OUTPUT, DriveCal.DRIVING_MAX_OUTPUT));
+
+    errors +=
+        SparkMaxUtils.check(
+            drivingSparkMax.setIdleMode(DriveConstants.DRIVING_MOTOR_IDLE_MODE));
+    errors +=
+        SparkMaxUtils.check(
+            drivingSparkMax.setSmartCurrentLimit(
+                DriveConstants.DRIVING_MOTOR_CURRENT_LIMIT_AMPS));*/ //TODO I cannot find a Talon equivlent for this stuff
```
There should be an equivalent to these two. Maybe not the output range one

voidSevenSevenSix replied:
fixed in configurator

balghane commented on src/main/java/frc/robot/subsystems/drive/SwerveModule.java:
```diff
@@ -0,0 +1,237 @@
+    // Apply chassis angular offset to the encoder position to get the position
+    // relative to the chassis.
+    return new SwerveModuleState(
+        drivingTalon.getVelocity().getValue()*DriveConstants.DRIVING_ENCODER_POSITION_FACTOR_METERS,
```
Are we sure on units here

voidSevenSevenSix replied:
fixed

balghane commented on src/main/java/frc/robot/subsystems/drive/SwerveModule.java:
```diff
@@ -0,0 +1,237 @@
+    this.desiredState = optimizedDesiredState;
+
+    // Command driving and turning SPARKS MAX towards their respective setpoints.
+    drivingTalon.setControl(
+        new DutyCycleOut(optimizedDesiredState.speedMetersPerSecond));
```
`DutyCycleOut` doesn't seem right for velocity control?

voidSevenSevenSix replied:
now using VelocityVoltage

balghane commented on src/main/java/frc/robot/subsystems/drive/SwerveModule.java:
```diff
@@ -0,0 +1,237 @@
+  }
+
+  public void initSendable(SendableBuilder builder) {
+    builder.addDoubleProperty("Driving kP", ()->{return drivingTalon.getClosedLoopProportionalOutput().getValue();}, null); //TODO find a setter? i couldn't
+    builder.addDoubleProperty("Driving kI", ()->{return drivingTalon.getClosedLoopIntegratedOutput().getValue();}, null);
+    builder.addDoubleProperty("Driving kD", ()->{return drivingTalon.getClosedLoopDerivativeOutput().getValue();}, null);
```
I think this is the contribution to the output from the P term, not the constant?

voidSevenSevenSix replied:
fixed, the above values are now marked as their respective contribution to output from the different terms

voidSevenSevenSix replied:
and the constants are also outputted to the builder

balghane commented on src/main/java/frc/robot/subsystems/drive/SwerveModule.java:
```diff
@@ -0,0 +1,237 @@
+    builder.addDoubleProperty("Turning kD", turningPIDController::getD, turningPIDController::setD);
+    builder.addDoubleProperty(
+        "Turning kFF", turningPIDController::getFF, turningPIDController::setFF);
+    builder.addDoubleProperty("Driving Vel (m/s)", ()->{ return drivingTalon.getVelocity().getValue();} , null);
```
Again are we sure on units here?

voidSevenSevenSix replied:
fixed

itsmesp8343 reviewed (commented):
👍 I asked James some questions and now for the most part I understand this.

tails618 committed:
Update Constants.java, RobotMap.java, DriveConstants.java, and DriveSubsystem.java

- Updated the comments in Constants.java to improve clarity.
- Changed the comment in RobotMap.java from "Driving SPARK MAX CAN IDs" to "Driving TalonFX CAN IDs".
- Removed unused imports in DriveConstants.java.
- Replaced placeholder values with actual values in DriveConstants.java.
- Removed unused variables and filters in DriveSubsystem.java.
- Updated the initialization of SwerveDriveOdometry in DriveSubsystem.java.

voidSevenSevenSix committed:
basel's fixes - someone fix my sensortomechanismratio math

voidSevenSevenSix committed:
Basel Fixes II

voidSevenSevenSix committed:
Basel Fixes II

voidSevenSevenSix committed:
Basel Fixes III

avakimc committed:
Revising PathPlanner Code in DriveSubsystem

- started using actual current module speeds
- changed it to total robot translation/rotation PID
- added TODO comments
- removed unnecessary variables/code
- make a path from not-yet smoothed points

avakimc committed:
Merge branch 'asher-saanika/swerve' of https://github.com/cachemoney8096/2024-crescendo into asher-saanika/swerve

tails618 committed:
feat: Add gyro configuration and apply mount pose

This commit adds the necessary code to configure the gyro and apply the mount pose. It imports the required classes from `com.ctre.phoenix6.configs` and `com.ctre.phoenix6.hardware`. The gyro is reset, and the mount pose values are set to 0 for pitch and roll, and 90 for yaw. Additionally, a comment suggests putting a certain block of code in a separate thread for faster looping.

The commit also removes some unused methods related to setting swerve module states, zeroing encoder positions, and initializing sendables.

Lastly, there is a TODO comment suggesting testing an arbitrary change in `LOOP_TIME_S` value to see if it improves performance.

Note: This commit message exceeds 50 characters but follows all other requirements specified by the user.

tails618 commented:
/format

github-actions[bot] committed:
Formatting fixes
